modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,733 @@
+programs such as this:
+
+.. code-block:: chapel
+
+  var A = Matrix(4, 4),
+      B = Matrix(4, 4);
+  // A + B is a promoted operation, and C becomes a 1D array
+  var C = A + B;
+  // This code would then result in an error due to rank mismatch:
+  C += A;
+
+To get avoid this, you can avoid relying on inferred-types for new arrays:
+
+.. code-block:: chapel
+
+  var A = Matrix(4, 4),
+      B = Matrix(4, 4);
+  // C's type is not inferred and promotion flattening is not a problem
+  var C: [A.domain] A.eltType = A + B;
+  // Works as expected
+  C += A;
+
+Alternatively, you can use the LinearAlgebra helper routines, which create
+and return a new array:
+
+.. code-block:: chapel
+
+  var A = Matrix(4, 4),
+      B = Matrix(4, 4);
+  // matPlus will create a new array with an explicit type
+  var C = matPlus(A, B);
+  // Works as expected
+  C += A;
+
+Promotion flattening is not expected to be an issue in future releases.
+
+
+.. LinearAlgebra Module Future TODOs:
+  - Support non-matching domains and eltTypes in matOperations
+    - check that eltTypes are coercible
+    - check domain shapes are equal
+  - Add LinearAlgebra primer
+  - Add feature table, comparing to numpy and matlab
+  - performance testing
+  - Domain maps
+    - Distributed array support
+    - Sparse support
+    - GPU support
+  - Support fully native implementations
+    - Only use the BLAS/LAPACK version when the libraries are installed
+  - Provide MKL BLAS/LAPACK with Chapel installation
+    - install with something like: CHPL_EXTRAS=MKL
+
+*/
+
+module LinearAlgebra {
+
+ use Norm; // TODO -- merge Norm into LinearAlgebra
+ use BLAS;
+// use LAPACK; // TODO -- Use LAPACK routines
+
+
+//
+// Matrix and Vector Initializers
+//
+
+/* Return a vector (1D array) over domain ``{0..#length}``*/
+proc Vector(length, type eltType=real) {
+  var V: [0..#length] eltType;
+  return V;
+}
+
+
+/* Return a vector (1D array) over domain ``Dom`` */
+proc Vector(Dom: domain(1), type eltType=real) {
+  var V: [Dom] eltType;
+  return V;
+}
+
+
+/* Return a vector (1D array) with domain and values of ``A`` */
+proc Vector(A: [?Dom] ?Atype, type eltType=Atype ) {
+  var V: [Dom] eltType = A;
+  return V;
+}
+
+
+pragma "no doc"
+proc Vector(x: ?t, Scalars...?n)  where isNumericType(t) {
+  type eltType = Scalars(1).type;
+  return Vector(x, (...Scalars), eltType=eltType);
+}
+
+
+/* Return a vector (1D array), given 2 or more numeric values
+
+   If `type` is omitted, it will be inferred from the first argument
+*/
+proc Vector(x: ?t, Scalars...?n, type eltType) where isNumericType(t) {
+
+  if !isNumeric(Scalars(1)) then
+      compilerError("Vector() expected numeric arguments");
+
+  var V: [0..n] eltType;
+
+  V[0] = x: eltType;
+
+  for i in 1..n do V[i] = Scalars[i]: eltType;
+
+  return V;
+}
+
+
+/* Return a matrix (2D array) over domain ``{0..#rows, 0..#cols}``*/
+proc Matrix(rows, cols, type eltType=real) where isIntegral(rows) && isIntegral(cols) {
```

Comment:
Also, I wonder if there should be a form of Matrix that just takes rows and returns a square matrix? That might be a trivial change to the above....